src/matchers/text.rs

```diff
@@ -0,0 +1,91 @@
+/// Returns whether a buffer is html data.
+///
+/// Conforms to [whatwg](https://mimesniff.spec.whatwg.org/)
+/// specification.
+pub fn is_html(buf: &[u8]) -> bool {
+    let values: &[&[u8]] = &[
+        b"<!DOCTYPE HTML",
+        b"<HTML",
+        b"<HEAD",
+        b"<SCRIPT",
+        b"<IFRAME",
+        b"<H1",
+        b"<DIV",
+        b"<FONT",
+        b"<TABLE",
+        b"<A",
+        b"<STYLE",
+        b"<TITLE",
+        b"<B",
+        b"<BODY",
+        b"<BR",
+        b"<P",
+        b"<!--",
+    ];
+    let buf = trim_start_whitespaces(buf);
+
+    for val in values {
+        if starts_with_ignore_ascii_case(buf, val) && buf.len() > val.len() {
+            match buf[val.len()] {
+                // tag-terminitating byte
+                0x20 | 0x3E => return true,
+                _ => continue,
+            }
+        }
+    }
+
+    false
+}
+
+/// Returns whether a buffer is xml data.
+///
+/// Conforms to [whatwg](https://mimesniff.spec.whatwg.org/)
+/// specification.
+pub fn is_xml(buf: &[u8]) -> bool {
+    let val: &[u8] = b"<?xml";
+    let buf = trim_start_whitespaces(buf);
+    starts_with_ignore_ascii_case(buf, val)
+}
+
+/// Strip whitespaces at the beginning of the buffer.
+///
+/// Follows https://mimesniff.spec.whatwg.org
+/// definition of whitespace.
+fn trim_start_whitespaces(buf: &[u8]) -> &[u8] {
+    for (i, b) in buf.iter().enumerate() {
+        match b {
+            0x09 | 0x0A | 0x0C | 0x0D | 0x20 => continue,
+            _ => return &buf[i..],
+        }
+    }
+    &[]
+}
```

Comment:
This could be rewritten this way, making the compiler optimize away a panic were `&buf[i..]` could be an invalid index  (which I don't think could ever happen) and almost halving the size of the `trim_start_whitespaces` assembly.

This makes it about 30x faster on my benchmark (around 20,000 ns/iter vs 800,000 ns/iter) on 1 MB of `0x09`. The previous version wasn't too bad, it was still under 1 ms/MB, but would have gotten worse on bigger files)

```suggestion
fn trim_start_whitespaces(mut buf: &[u8]) -> &[u8] {
    while !buf.is_empty() {
        match buf[0] {
            0x09 | 0x0A | 0x0C | 0x0D | 0x20 => buf = &buf[1..],
            _ => return buf,
        }
    }
    &[]
}
```

You can try the benchmark yourself by doing:

```rust
// lib.rs
#![feature(test)]

// inside mod tests { in your file
    extern crate test;
    use test::Bencher;

    #[bench]
    fn bench_whitespace(b: &mut Bencher) {
        b.iter(|| {
            let mut v = vec![0x09; 1000000];
            v.push(b'A');

            let n = test::black_box(v);
            trim_start_whitespaces(&n);
        })
    }
```

Reply:
On Windows 10 I do not have the same results:

```
# without optimization
test matchers::text::tests::bench_whitespace ... bench:     957,211 ns/iter (+/- 63,163)
# with optimization
test matchers::text::tests::bench_whitespace ... bench:     428,169 ns/iter (+/- 57,372)
```

I will test on Linux tonight.
I am a bit disappointed by Rust compiler/LLVM here. I though that iterator abstractions would be transcoded to assembly that I could not hand write better. When you look at your code and mine, the compiler should abstract that away, isn't it ?

Reply:
Speaking at optimizations have you looked at optimizing the global match algorithm ? Instead of iterating through all matchers I was thinking that building a (static) binary tree for each byte might be better, but it is pure intuition, I have no implementation to verify that.

Reply:
That's very strange. It generated very similar assembly with the windows gnu target, but very bad assembly with the msvc target.

While looking at the function I realized I could make it even simpler by breaking when a non space character is reached and now the assembly is the same.

## Before

```rust
fn trim_start_whitespaces(mut buf: &[u8]) -> &[u8] {
    while !buf.is_empty() {
        match buf[0] {
            0x09 | 0x0A | 0x0C | 0x0D | 0x20 => buf = &buf[1..],
            _ => return buf,
        }
    }
    &[]
}
```

## After
```rust
fn trim_start_whitespaces(mut buf: &[u8]) -> &[u8] {
    while !buf.is_empty() {
        match buf[0] {
            0x09 | 0x0A | 0x0C | 0x0D | 0x20 => buf = &buf[1..],
            _ => break,
        }
    }
    buf
}
```

I updated the code suggestion

Reply:
This does not change the results on nightly-x86_64-pc-windows-msvc.

Reply:
I don't have a windows computer to test this on, so I'm not sure what's happening here. I could only guess compiling on Linux with a Windows target generates different assembly from compiling natively on Windows?

Reply:
I will try to reproduce on Linux tonight and will come back to you with my findings.
After that you might consider opening an issue on Rust project directly if you like such topic: I believe the compiler should optimize my code and yours the same way. The performance gain on Linux (which I will try to reproduce tonight) is too big to be ignored here.

Reply:
In your implementation the biggest issue is that it's not able to elide the panic. This makes it almost as fast as mine (except on Windows msvc because of that `&[]`)

```rust
fn trim_start_whitespaces(buf: &[u8]) -> &[u8] {
    for (i, b) in buf.iter().enumerate() {
        match b {
            0x09 | 0x0A | 0x0C | 0x0D | 0x20 => continue,
            _ => return unsafe { buf.get_unchecked(i..) },
        }
    }
    &[]
}
```

I'll file an issue later

Reply:
Because of https://github.com/rust-lang/rust/pull/74510 the panic might not be a `slice_index_order_fail`, but a bounds check, since this generates the same result

```rust
fn trim_start_whitespaces(buf: &[u8]) -> &[u8] {
    for (i, b) in buf.iter().enumerate() {
        match b {
            0x09 | 0x0A | 0x0C | 0x0D | 0x20 => continue,
            // incrementing i by 1 so now we can't be sure we are in range
            _ => return &buf[i+1..],
        }
    }
    &[]
}
```

Reply:
Just to understand: `&buf[1..]` is not handled the same as `&buf[i..]` ? I do not get the difference between the 2 codes actually.

Reply:
In the first case I think it knowns `buf.len() > 0` so `&buf[1..]` will never fail. In the second case it probably doesn't understand that` i < buf.len()`. The same happens if you remove enumerate and just increment `i` youself after each iteration

Reply:
I found something very interesting while trying rewriting this function in many many different ways. I opened an issue

https://github.com/rust-lang/rust/issues/74688

Reply:
I could not imagine that such innocent PR would lead to a bug opened to Rust's compiler. Well done !